postman_collection.json

```diff
@@ -1,2124 +1,3260 @@
ansportCallSequenceNumber\": 1,\n                \"modeOfTransport\": \"RAIL\",\n                \"location\": {\n                    \"UNLocationCode\": \"USMIA\",\n                    \"locationName\": \"Nijmegen\"\n                },\n                \"facilityTypeCode\": \"INTE\"\n            }\n        }\n    },\n    {\n        \"metadata\": {\n            \"eventID\": \"2968b966-ee81-46ba-af87-0c5031c64142\",\n            \"eventCreatedDateTime\": \"2021-11-28T13:12:56Z\",\n            \"eventType\": \"TRANSPORT\"\n        },\n        \"payload\": {\n            \"eventDateTime\": \"2021-11-30T23:11:00Z\",\n            \"eventClassifierCode\": \"PLN\",\n            \"relatedDocumentReferences\": [\n                {\n                    \"type\": \"CBR\",\n                    \"value\": \"CARRIER_BOOKING_REQUEST_REFERENCE_02\"\n                },\n                {\n                    \"type\": \"BKG\",\n                    \"value\": \"ABC123123123\"\n                }\n            ],\n            \"transportEventTypeCode\": \"DEPA\",\n            \"delayReasonCode\": \"WEA\",\n            \"changeRemark\": \"Bad weather\",\n            \"transportCall\": {\n                \"transportCallReference\": \"TC-REF-08_02-A\",\n                \"transportCallSequenceNumber\": 1,\n                \"modeOfTransport\": \"VESSEL\",\n                \"location\": {\n                    \"UNLocationCode\": \"USNYC\",\n                    \"locationName\": \"Eiffel Tower\",\n                    \"latitude\": \"48.8585500\",\n                    \"longitude\": \"2.294492036\"\n                },\n                \"facilityTypeCode\": \"POTE\",\n                \"vessel\": {\n                    \"vesselIMONumber\": \"1234567\",\n                    \"name\": \"King of the Seas\",\n                    \"flag\": \"DE\",\n                    \"callSign\": \"NCVV\",\n                    \"operatorCarrierCode\": \"MSK\",\n                    \"operatorCarrierCodeListProvider\": \"SMDG\"\n                }\n            }\n        }\n    },\n    {\n        \"metadata\": {\n            \"eventID\": \"9d5d0824-b228-4ea8-b2cb-4ebd8da76e15\",\n            \"eventCreatedDateTime\": \"2021-11-29T13:12:56Z\",\n            \"eventType\": \"TRANSPORT\"\n        },\n        \"payload\": {\n            \"eventDateTime\": \"2021-12-02T23:11:00Z\",\n            \"eventClassifierCode\": \"PLN\",\n            \"relatedDocumentReferences\": [\n                {\n                    \"type\": \"CBR\",\n                    \"value\": \"a3a34f10-acc5-4e23-b52e-146f63458c90\"\n                },\n                {\n                    \"type\": \"CBR\",\n                    \"value\": \"ef223019-ff16-4870-be69-9dbaaaae9b11\"\n                },\n                {\n                    \"type\": \"BKG\",\n                    \"value\": \"C501576CD94F\"\n                },\n                {\n                    \"type\": \"BKG\",\n                    \"value\": \"DCR987876762\"\n 
```

Comment:
From this line 3217, many variables are added for the CTK. Those are used mainly for the pre-test script and a few additional tests for the CTK.

Reply:
It is hard to review such a huge JSON file. Can u pls load it in Postman for review? You will see many pre-tests script and additional test such as 


Reference implementation request body mandatory field check
PASS

Send a HEAD request to the given callback URL https://f12d95eb-6e7b-4516-ac10-7c3c9efe2887.mock.pstmn.io/456eacf9-8cda-412b-b801-4a41be7a6d000
PASS
Send a POST request to the given callback URL https://f12d95eb-6e7b-4516-ac10-7c3c9efe2887.mock.pstmn.io/456eacf9-8cda-412b-b801-4a41be7a6d000
PASS
Check response schema
PASS
Check request body schema

Reply:
I copy the code for the pre-test script here:

// Additional test for conformance
const testMode = pm.collectionVariables.get("TEST_MODE");
const conformanceTest = pm.collectionVariables.get("CONFORMANCE_TEST_MODE");
var requestBody = pm.collectionVariables.get("CONFORMANCE_EVENT_SUBSCRIPTION_POST_REQUEST_BODY");

if(testMode.includes(conformanceTest) ){
    pm.test("conformance request body mandatory field check", function () {
        try {
            var parsedBody = JSON.parse(requestBody);
            if (!parsedBody.secret) {
                throw new Error("Required field secret is missing");
            }
            if(!parsedBody.callbackUrl){
                throw new Error("Required field callbackUrl is missing.");
            }    
    } catch (error) {
            pm.test("Request Body Assertion", function () {
                pm.expect.fail("Failed to parse request body or missing required fields: " + error.message);
            });
    }
});
    pm.request.body = requestBody;
    pm.request.headers.add({key: "Content-Type", value: "application/json"});
}else{
    pm.test("Reference implementation request body mandatory field check", function () {
        const requiredFields = ['callbackUrl', 'secret'];
        requestBody = JSON.parse(pm.request.body.raw);
        requiredFields.forEach(field => {
            if (!requestBody.hasOwnProperty(field)) {
                throw new Error(`Missing ${field} field in the request body`);
            }
        });
    });
}

Reply:
Also extra test for the CTK:

//Postman mock server for subsription callback
const reqBody = JSON.parse(request.data);

pm.test("Send a HEAD request to the given callback URL "+reqBody.callbackUrl, function () {
    // Define the request options
    const requestOptions = {
        method: 'HEAD',
        url: reqBody.callbackUrl,
        header: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
        json: true
    };

    // Send the request 
    pm.sendRequest(requestOptions, function (error, response) {
        // Handle the response
        if (error) {
            console.log(error);
            throw new Error("HEAD request error for the callbackUrl "+error);
        } else {
           pm.expect(response).to.have.property('code', 204);
        }
    });
});

pm.test("Send a POST request to the given callback URL "+reqBody.callbackUrl, function () {
    // Define the request options
    const requestOptions = {
        method: 'POST',
        url: reqBody.callbackUrl,
        header: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
        json: true
    };

    // Send the request 
    pm.sendRequest(requestOptions, function (error, response) {
        // Handle the response
        if (error) {
            console.log(error);
            throw new Error("POST request error for the callbackUrl "+error);
        } else {
           pm.expect(response).to.have.property('code', 204);
        }
    });
});



// Additional test for conformance
const response_schema = pm.collectionVariables.get("EVENT_SUBSCRIPTION_RESPONSE_SCHEMA");
const request_schema = pm.collectionVariables.get("EVENT_SUBSCRIPTION_REQUEST_SCHEMA");
const testMode = pm.collectionVariables.get("TEST_MODE");
const ignoreSchemaTest = pm.collectionVariables.get("IGNORE_SCHEMA_TEST_MODE");
const conformanceTest = pm.collectionVariables.get("CONFORMANCE_TEST_MODE");
const and = pm.collectionVariables.get("AND");


if( !testMode.includes(ignoreSchemaTest) ){
    pm.test("Check response schema", function(){
        var result = tv4.validateResult(pm.response.json(), response_schema);
        if(!result.valid){
            console.log(`Schema validation failed: ${JSON.stringify(validationResult.errors)}`);
            console.log(`Response: ${JSON.stringify(response)}`);
            throw new Error("Response schema validation failed");
        }
        pm.expect(result.valid).to.be.true;
    });
    
    pm.test("Check request body schema", function(){
        var requestData;    
        if(testMode === conformanceTest){
            requestData = pm.collectionVariables.get("CONFORMANCE_EVENT_SUBSCRIPTION_POST_REQUEST_BODY");
        }else{
            requestData = JSON.parse(pm.request.body.raw);
        }    
        const validationResult = tv4.validateResult(requestData, request_schema);
        if (!validationResult.valid) {
            console.log(`Schema validation failed: ${JSON.stringify(validationResult.errors)}`);
            console.log(`Response: ${JSON.stringify(response)}`);
            throw new Error("Schema validation failed "+validationResult.errors);
        }
        pm.expect(validationResult.valid).to.be.true;
    });
}

Reply:
Note I didn't add any new Postman requests for this PR. Just additional pre-test script and postman test.

Reply:
I see. I did import it into postman but only saw the same requests, I did not look at the tests for each request as I assumed the CTK would have it's own set of requests.